llvm/lib/CodeGen/MacroFusion.cpp

```diff
@@ -137,19 +137,33 @@ namespace {
 /// Post-process the DAG to create cluster edges between instrs that may
 /// be fused by the processor into a single operation.
 class MacroFusion : public ScheduleDAGMutation {
-  ShouldSchedulePredTy shouldScheduleAdjacent;
+  std::vector<MacroFusionPredTy> Predicates;
   bool FuseBlock;
   bool scheduleAdjacentImpl(ScheduleDAGInstrs &DAG, SUnit &AnchorSU);
 
 public:
-  MacroFusion(ShouldSchedulePredTy shouldScheduleAdjacent, bool FuseBlock)
-    : shouldScheduleAdjacent(shouldScheduleAdjacent), FuseBlock(FuseBlock) {}
+  MacroFusion(std::vector<MacroFusionPredTy> Predicates, bool FuseBlock)
```

Comment:
Maybe this should be an ArrayRef<MacroFusionPredTy> and only copied once we get here?

Reply:
I used `ArrayRef` here at the first time, but we got `invalid permissions for mapped object` errors. It seems that the `std::vector` that `ArrayRef` refers to has been destroyed.
(I am really not fimilar with C++ object lifetime 🤕)